### PR TITLE
Jp-1273 Implement mrs residual fringe

### LINKED ISSUE #4531: Residual fringe correction needs an algorithm
Issue [JP-1273](https://jira.stsci.edu/browse/JP-1273) was created by Alicia Canipe:

From [~kgordon]:
 * Residual fringe correction needs an algorithm

### PR DESCRIPTION
**Pull request by jemorrison**
This PR addresses [JP-1273](https://jira.stsci.edu/browse/JP-1273): implement a MIRI MRS residual fringe correction
Closes #4531 

The algorithm was supplied by the MIRI EC instrument team.  A third party fitting routine is used and is
included in the PR.   It would be good if we can remove this third party fitting library and use astropy
instead.  This PR tried to copy the original algorithm (as much as possible).
The algorithm loops over every slice on the IFU data and tries to fit each column in the slice. The algorithm first estimates the number of fringes in each column. If not enough estimated fringes are found the code skips trying to correct for a fringe for that column. There are several other reasons the fitting is skipped. In this PR information is printed to screen if the fringe fitting is skipped for a column,. 

Data for testing this (along with a residual fringe reference file) is found at:
/grp/jwst/ssb/morrison/MIRI_MRS_FRINGEDATA/

Checklist
- [ ] Tests
- [ ] Documentation
- [ ] Change log
- [x] Milestone
- [x] Label(s)

### CODE REVIEW

**Comment by jemorrison** [comment]
@nden  I sort of assumed we would be making changes to the third party packages. There are many many flake8 errors with all the 3rd party software I pulled in. If we are going to merge this PR then I guess I need to fix all those flake 8 errors

**Comment by nden** [comment]
@jemorrison Try skipping the 3rd party file checks by adding 
# flake8: noqa
at the top of the files.

**Comment by jemorrison** [comment]
@howard I have not set up any tests. I thought I should wait and see what fitting package we used. But if you think I should put some in I can

**Comment by jemorrison** [comment]
@eslavich  do you have an idea why some of these CI tests are failing

**Comment by eslavich** [comment]
The new step(s) need to be added to this list:

https://github.com/spacetelescope/jwst/blob/master/jwst/stpipe/integration.py#L23

File: jwst/lib/suffix.py [hunk 1]
```diff
@@ -103,6 +103,8 @@
     'skymatch',
     'extract_2d',
     'cubebuildstep',
+    'residualfringstep',
```
Comment (eslavich) [inline jwst/lib/suffix.py:106]: There's a typo here, missing an `e`

File: jwst/lib/suffix.py [hunk 2]
```diff
@@ -103,6 +103,8 @@
...
+    'residual_fringe',
```
Comment (eslavich) [inline jwst/lib/suffix.py:107]: The presence of this value may be causing the test to fail (but I'm not familiar with this suffix code, so maybe not).

**Comment by jemorrison** [comment]
@stscieisenhamer the CI failures seem to be related to this routine. I am missing something adding a new step but I don't know what it is. Do you know ? 
jwst/lib/tests/test_suffix.py::test_suffix_existence - AssertionError:..

**Comment by jemorrison** [comment]
@hbushouse @stscieisenhamer  I believe this is ready to be merged

File: jwst/lib/suffix.py [hunk 3]
```diff
@@ -46,7 +46,7 @@
     'ramp', 'rate', 'rateints',
     's2d', 's3d', 'snr',
     'uncal',
-    'wfscmb', 'whtlt',
+    'wfscmb', 'whtlt','residual_fringe',
```
Comment (stscieisenhamer) [inline jwst/lib/suffix.py:49]: Minor ADHD nit: Keeping this list alphabetical will help track down future issues 😄

File: jwst/residual_fringe/residual_fringe_step.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        save_intermediate_results  = boolean(default = False)
+        transmission_level = integer(default=80) # transmission level to use to define slice locations
+        search_output_file = boolean(default = False)
+    """
+
+    reference_file_types = ['fringefreq','regions']
+
+    def process(self, input):
+
+        self.suffix = 'residual_fringe'
```
Comment (stscieisenhamer) [inline jwst/residual_fringe/residual_fringe_step.py:31]: This should be done in the `spec`. Add the following into the spec:

`suffix = string(default="residual_fringe")`

**Review by stscieisenhamer — APPROVED** [review]
Just some minor potential updates, but otherwise LGTM

**Pull request merged**